File: packages/game-lib/src/lib/behaviors/platformer-3d/components.ts
/**
 * Platformer 3D ECS Components
 * 
 * Components for 3D platformer movement system including walking, running,
 * jumping with multi-jump support, and falling/landing states.
 */

// ─────────────────────────────────────────────────────────────────────────────
// Platformer3DMovementComponent (capability / configuration)
// Defines the movement capabilities and physics parameters
// ─────────────────────────────────────────────────────────────────────────────

export interface Platformer3DMovementComponent {
	/** Base walking speed */
	walkSpeed: number;
	/** Sprint/run speed */
	runSpeed: number;
	/** Initial jump force/velocity */
	jumpForce: number;
	/** Maximum number of jumps (1 = single jump, 2 = double jump, etc.) */
	maxJumps: number;
	/** Gravity force (optional override) */
	gravity: number;
	/** Ray length for ground detection */
	groundRayLength: number;
	/** Coyote time in seconds (grace period after leaving ground) */
	coyoteTime: number;
	/** Jump buffer time in seconds (queue jump input before landing) */
	jumpBufferTime: number;
	/** Velocity multiplier when releasing jump button early (0-1) */
	jumpCutMultiplier: number;
}

export function createPlatformer3DMovementComponent(
	options: Partial<Platformer3DMovementComponent> = {}
): Platformer3DMovementComponent {
	return {
		walkSpeed: options.walkSpeed ?? 12,
		runSpeed: options.runSpeed ?? 24,
		jumpForce: options.jumpForce ?? 12,
		maxJumps: options.maxJumps ?? 1,
		gravity: options.gravity ?? 9.82,
		groundRayLength: options.groundRayLength ?? 1.0,
		coyoteTime: options.coyoteTime ?? 0.1,
		jumpBufferTime: options.jumpBufferTime ?? 0.1,
		jumpCutMultiplier: options.jumpCutMultiplier ?? 0.5,
	};
}

// ─────────────────────────────────────────────────────────────────────────────
// Platformer3DInputComponent (intent)
// Written by: Player controller, AI controller
// Read by: Platformer3DBehavior
// ─────────────────────────────────────────────────────────────────────────────

export interface Platformer3DInputComponent {
	/** Horizontal movement input (-1 to 1) */
	moveX: number;
	/** Forward/backward movement input (-1 to 1) */
	moveZ: number;
	/** Jump button pressed */
	jump: boolean;
	/** Run/sprint button held */
	run: boolean;
}

export function createPlatformer3DInputComponent(): Platformer3DInputComponent {
	return {
		moveX: 0,
		moveZ: 0,
		jump: false,
		run: false,
	};
}

// ─────────────────────────────────────────────────────────────────────────────
// Platformer3DStateComponent (runtime state)
// Tracks the current state of the platformer entity
// ─────────────────────────────────────────────────────────────────────────────

export interface Platformer3DStateComponent {
	/** Is the entity on the ground */
	grounded: boolean;
	/** Is currently in a jump arc */
	jumping: boolean;
	/** Is falling (not jumping) */
	falling: boolean;
	/** Current jump count (for multi-jump) */
	jumpCount: number;
	/** Y position when jump started */
	jumpStartHeight: number;
	/** Current movement speed being applied */
	currentSpeed: number;
	/** Last known ground Y position */
	lastGroundedY: number;
	/** Was jump button pressed last frame (for edge detection) */
	jumpPressedLastFrame: boolean;
	
	// -- New fields --
	/** Is currently colliding with ground (from physics collision callback) */
	collisionGrounded: boolean;
	/** Timestamp of last ground collision */
	groundedCollisionTime: number;
	/** Time in seconds since last grounded */
	timeSinceGrounded: number;
	/** Is a jump allowed due to buffer? */
	jumpBuffered: boolean;
	/** Time remaining in jump buffer */
	jumpBufferTimer: number;
	/** Is jump button currently held */
	jumpHeld: boolean;
	/** Has the jump cut been applied for this jump? */
	jumpCutApplied: boolean;
}

export function createPlatformer3DStateComponent(): Platformer3DStateComponent {
	return {
		grounded: false,
		jumping: false,
		falling: false,
		jumpCount: 0,
		jumpStartHeight: 0,
		currentSpeed: 0,
		lastGroundedY: 0,
		jumpPressedLastFrame: false,
		collisionGrounded: false,
		groundedCollisionTime: 0,
		timeSinceGrounded: 0,
		jumpBuffered: false,
		jumpBufferTimer: 0,
		jumpHeld: false,
		jumpCutApplied: false,
	};
}

File: packages/zylem-runtime/src/runtime/behaviors/platformer_3d/components.rs
//! Components for the 3D platformer behavior.
//!
//! These mirror the TypeScript `platformer-3d` components in
//! `packages/game-lib/src/lib/behaviors/platformer-3d/components.ts` but live
//! on the wasm runtime side and operate on `Body3DSlot` poses driven by
//! Rapier's `KinematicCharacterController`.

/// Tunable platformer movement parameters.
///
/// Defaults are intentionally identical to the TypeScript
/// `createPlatformer3DMovementComponent` factory so behavior stays consistent
/// when host code switches from the TS implementation to the wasm one.
#[derive(Clone, Copy, Debug)]
pub struct Platformer3DConfig {
    pub walk_speed: f32,
    pub run_speed: f32,
    pub jump_force: f32,
    pub max_jumps: u32,
    /// Magnitude of downward gravity applied to `linvel.y` when airborne.
    /// Stored as a positive number; the sign is applied internally.
    pub gravity: f32,
    pub coyote_time: f32,
    pub jump_buffer_time: f32,
    pub jump_cut_multiplier: f32,
    pub multi_jump_window: f32,
    /// Maximum slope angle (in degrees) the controller can climb. Forwarded
    /// to KCC's `max_slope_climb_angle` at bootstrap.
    pub max_slope_deg: f32,
    /// Auto-step height (world units). 0 disables stairs/curb climbing.
    pub autostep_height: f32,
    /// Snap-to-ground length (world units). 0 disables ground snapping.
    pub snap_to_ground: f32,
}

impl Default for Platformer3DConfig {
    fn default() -> Self {
        Self {
            walk_speed: 12.0,
            run_speed: 24.0,
            jump_force: 12.0,
            max_jumps: 1,
            gravity: 9.82,
            coyote_time: 0.1,
            jump_buffer_time: 0.1,
            jump_cut_multiplier: 0.5,
            multi_jump_window: 0.15,
            max_slope_deg: 50.0,
            autostep_height: 0.3,
            snap_to_ground: 0.2,
        }
    }
}

/// Per-frame platformer intent. Written by host code each tick before `step`.
#[derive(Clone, Copy, Debug, Default)]
pub struct Platformer3DInput {
    pub move_x: f32,
    pub move_z: f32,
    pub jump: bool,
    pub run: bool,
}

/// Runtime state tracking jump FSM, coyote/buffer windows, and grounded
/// timing. Mirrors the TypeScript `Platformer3DStateComponent` shape.
#[derive(Clone, Copy, Debug)]
pub struct Platformer3DState {
    pub grounded: bool,
    pub jumping: bool,
    pub falling: bool,
    pub jump_count: u32,
    pub jump_pressed_last_frame: bool,
    pub jump_held: bool,
    /// Must release jump button before each multi-jump.
    pub jump_released_since_last_jump: bool,
    pub jump_buffered: bool,
    pub jump_buffer_timer: f32,
    pub jump_cut_applied: bool,
    /// Remaining seconds of jump-cut velocity ramp (lerps `vy` toward
    /// `vy * jump_cut_multiplier` to avoid a single-frame discontinuity).
    pub jump_cut_ramp_timer: f32,
    pub time_since_grounded: f32,
    pub time_since_jump: f32,
    /// Set true on the frame a jump impulse is applied. Suppresses gravity
    /// for exactly that frame so the jump impulse is the authoritative `vy`.
    pub jumped_this_frame: bool,
    pub current_speed: f32,
    pub last_grounded_y: f32,
}

impl Default for Platformer3DState {
    fn default() -> Self {
        Self {
            grounded: false,
            jumping: false,
            falling: false,
            jump_count: 0,
            jump_pressed_last_frame: false,
            jump_held: false,
            jump_released_since_last_jump: true,
            jump_buffered: false,
            jump_buffer_timer: 0.0,
            jump_cut_applied: false,
            jump_cut_ramp_timer: 0.0,
            time_since_grounded: 0.0,
            time_since_jump: 0.0,
            jumped_this_frame: false,
            current_speed: 0.0,
            last_grounded_y: 0.0,
        }
    }
}

/// Duration of the jump-cut velocity ramp. Lifted directly from
/// `JUMP_CUT_RAMP_DURATION` in the TS implementation so behavior matches.
pub const JUMP_CUT_RAMP_DURATION: f32 = 0.05;
